File: id/id-oauth-google/src/main/java/io/piano/android/id/google/GoogleSignInActivity.java
package io.piano.android.id.google;

import android.app.Activity;
import android.content.Intent;
import android.os.Bundle;

import androidx.annotation.Nullable;

import com.google.android.gms.auth.api.signin.GoogleSignIn;
import com.google.android.gms.auth.api.signin.GoogleSignInAccount;
import com.google.android.gms.auth.api.signin.GoogleSignInClient;
import com.google.android.gms.auth.api.signin.GoogleSignInOptions;
import com.google.android.gms.auth.api.signin.GoogleSignInStatusCodes;
import com.google.android.gms.common.api.ApiException;

import io.piano.android.id.PianoId;
import io.piano.android.id.PianoOAuthActivity;

public class GoogleSignInActivity extends PianoOAuthActivity {

    private static final int RC_SIGN_IN = 1;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        try {
            GoogleSignInOptions options = new GoogleSignInOptions.Builder(GoogleSignInOptions.DEFAULT_SIGN_IN)
                    .requestEmail()
                    .requestProfile()
                    .requestIdToken(getIntent().getStringExtra(PianoId.KEY_CLIENT_ID))
                    .build();

            GoogleSignInClient signInClient = GoogleSignIn.getClient(this, options);
            signInClient.signOut();

            Intent intent = signInClient.getSignInIntent();
            startActivityForResult(intent, RC_SIGN_IN);
        } catch (Exception e) {
            setFailureResult(e);
            finish();
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        if (requestCode == RC_SIGN_IN) {
            try {
                GoogleSignInAccount account = GoogleSignIn
                        .getSignedInAccountFromIntent(data)
                        .getResult(ApiException.class);

                setSuccessResult(GoogleOAuthProvider.NAME, account.getIdToken());
            } catch (ApiException e) {
                switch (e.getStatusCode()) {
                    case GoogleSignInStatusCodes.CANCELED:
                    case GoogleSignInStatusCodes.SIGN_IN_CANCELLED:
                        setResult(Activity.RESULT_CANCELED);
                        break;
                    default:
                        setFailureResult(e);
                }
            } catch (Exception e) {
                setFailureResult(e);
            }
        } else super.onActivityResult(requestCode, resultCode, data);

        finish();
    }
}

File: id/id-oauth-google/src/main/java/io/piano/android/id/google/GoogleSignInActivity.kt
package io.piano.android.id.google

import android.os.Bundle
import androidx.credentials.CredentialManager
import androidx.credentials.GetCredentialRequest
import androidx.credentials.exceptions.GetCredentialCancellationException
import androidx.lifecycle.lifecycleScope
import com.google.android.libraries.identity.googleid.GetSignInWithGoogleOption
import com.google.android.libraries.identity.googleid.GoogleIdTokenCredential
import io.piano.android.id.PianoId
import io.piano.android.id.PianoOAuthActivity
import kotlinx.coroutines.launch

public class GoogleSignInActivity : PianoOAuthActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (savedInstanceState == null) {
            try {
                val googleIdOption = GetSignInWithGoogleOption.Builder(requireNotNull(intent.getStringExtra(PianoId.KEY_CLIENT_ID)))
                    .build()
                val request = GetCredentialRequest(
                    credentialOptions = listOf(googleIdOption)
                )
                val credentialManager = CredentialManager.create(this)
                lifecycleScope.launch {
                    try {
                        val credential = credentialManager.getCredential(this@GoogleSignInActivity, request).credential
                        check(credential.type == GoogleIdTokenCredential.TYPE_GOOGLE_ID_TOKEN_CREDENTIAL) {
                            "Unsupported credential type '${credential.type}'"
                        }
                        val googleIdTokenCredential = GoogleIdTokenCredential.createFrom(credential.data)
                        setSuccessResult(GoogleOAuthProvider.NAME, googleIdTokenCredential.idToken)
                    } catch (e: GetCredentialCancellationException) {
                        setResult(RESULT_CANCELED)
                    } catch (e: Exception) {
                        setFailureResult(e)
                    } finally {
                        finish()
                    }
                }
            } catch (e: Exception) {
                setFailureResult(e)
                finish()
            }
        }
    }
}
